src/python/gudhi/wasserstein/wasserstein.py

```diff
@@ -88,32 +92,137 @@ def _perstot(X, order, internal_p, enable_autodiff):
+    :param axis: must be 0 or 1, correspond to the coordinate which is finite.
+    :returns: cost (float) and match for points with *one* infinite coordinate.
+
+    .. note::
+        Assume idX, idY come when calling _handle_essential_parts, thus have same length.
+    '''
+    u = X[idX, axis]
+    v = Y[idY, axis]
+
+    cost = np.sum(np.abs(np.sort(u) - np.sort(v))**(order))  # OT cost in 1D
+
+    sortidX = idX[np.argsort(u)]
+    sortidY = idY[np.argsort(v)]
+    # We return [i,j] sorted per value, and then [i, -1] (or [-1, j]) to account for essential points matched to the diagonal
+    match = list(zip(sortidX, sortidY))
+
+    return cost, match
+
+
+def _handle_essential_parts(X, Y, order):
+    '''
+    :param X: (n x 2) numpy array, first diagram.
+    :param Y: (n x 2) numpy array, second diagram.
+    :order: Wasserstein order for cost computation.
+    :returns: cost and matching due to essential parts. If cost is +inf, matching will be set to None.
+    '''
+    c = 0
+    m = []
+
+    ess_parts_X = _get_essential_parts(X)
+    ess_parts_Y = _get_essential_parts(Y)
+
+    # Treats the case of infinite cost (cardinalities of essential parts differ).
+    for u, v in zip(ess_parts_X, ess_parts_Y):
+        if len(u) != len(v):
+            return np.inf, None
+
+    # Now we know each essential part has the same number of points in both diagrams.
+    # Handle type 0 and type 1 essential parts (those with one finite coordinates)
+    c1, m1 = _cost_and_match_essential_parts(X, Y, ess_parts_X[0], ess_parts_Y[0], axis=1, order=order)
+    c2, m2 = _cost_and_match_essential_parts(X, Y, ess_parts_X[1], ess_parts_Y[1], axis=0, order=order)
+
+    c += c1 + c2
+    m += m1 + m2
+
+    # Handle type >= 2 (both coordinates are infinite, so we essentially just align points)
+    for u, v in zip(ess_parts_X[2:], ess_parts_Y[2:]):
+        m += list(zip(u, v))  # cost is 0
+
+    return c, np.array(m)
+
+
+def _offdiag(X, enable_autodiff):
+    '''
+    :param X: (n x 2) numpy array encoding a persistence diagram.
+    :param enable_autodiff: boolean, to handle the case where X is a eagerpy tensor.
+    :returns: The off-diagonal part of a diagram `X` (points with finite coordinates).
+    '''
+    if enable_autodiff:
+        # Assumes the diagrams only have finite coordinates. Thus, return X directly.
+        # TODO improve this to get rid of essential parts if there are any.
+        return X
+    else:
+        return X[np.where(np.isfinite(X[:,0]) & np.isfinite(X[:,1]))]
+
+
+def wasserstein_distance(X, Y, matching=False, order=1., internal_p=np.inf, enable_autodiff=False,
+                         keep_essential_parts=True):
     '''
-    :param X: (n x 2) numpy.array encoding the (finite points of the) first diagram. Must not contain essential points
-                (i.e. with infinite coordinate).
+    :param X: (n x 2) numpy.array encoding the first diagram. Can now contain essential parts (points with infinite
```

Comment:
I think we should avoid "now" which belongs in release notes and would soon be outdated in the main doc.